content/02_KeyTypes.md

```diff
@@ -1,488 +1,627 @@
+
+If you're wondering whether the relationship between `IObservable<T>` and `IObserver<T>` is similar to the relationship between [`IEnumerable<T>`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.ienumerable-1) and [`IEnumerator<T>`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.ienumerator-1), then you're onto something. Both `IEnumerator<T>` and `IObservable<T>` represent _potential_ sequences. With both of these interfaces, they will only supply data if we ask them for it. To get values out of an `IEnumerable<T>`, an `IEnumerator<T>` needs to come into existence, and similarly, to get values out of an `IObservable<T>` requires an `IObserver<T>`.
+
+The difference reflects the fundamental _pull vs push_ difference between `IEnumerable<T>` and `IObservable<T>`. Whereas with `IEnumerable<T>` we ask the source to create an `IEnumerator<T>` for us which we can then use to retrieve items (which is what a C# `foreach` loop does), with `IObservable<T>`, the source does not _implement_ `IObserver<T>`: it expects _us_ to supply an `IObserver<T>` and it will then push its values into that observer.
+
+So why does `IObserver<T>` have these three methods? Remember when I said that in an abstract sense, `IObserver<T>` represents the same thing as `IEnumerable<T>`? I meant it. It might be an abstract sense, but it is precise: `IObservable<T>` and `IObserver<T>` were designed to preserve the exact meaning of `IEnumerable<T>` and `IEnumerator<T>`, changing only the detailed mechanism of consumption.
+
+To see what that means, think about what happens when you iterate over an `IEnumerable<T>` (with, say, a `foreach` loop). With each iteration (and more precisely, on each call to the enumerator's [`MoveNext`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.ienumerator.movenext) method) there are three things that could happen:
+
+* `MoveNext` could return `true` to indicate that a value is available in the enumerator's [`Current`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.ienumerator-1.current) property
+* `MoveNext` could throw an exception
+* `MoveNext` could return `false` to indicate that you've reached the end of the collection
+
+These three outcomes correspond precisely to the three methods defined by `IObserver<T>`. We could describe these in slightly more abstract terms:
```

Comment:
I think there's an interesting point in there trying to get out, but don't think those are quite the right words.

Having watched most of the old Meijer, Beckman, and Bart videos from the dawn of Rx, this certainly doesn't seem to be how they thought about it. Their perspective, as I understand it, is that this split into three pieces is very much not at attempt to make Rx more pure; on the contrary, it's an attempt to make Rx exactly model what `IEnumerable<T>` is. It's a simple consequence of the duality: these three facets that were combined into `MoveNext`+`Current` become explicitly visible in `IObserver<T>`.

Eric and Bart have independently talked about what purified versions of `IEnumerable<T>` and `IObservable<T>` would look like. (They've both done videos on this, and Bart did a blog about it too.) They both show a simpler model where you don't do exceptions, and where you model the end of a sequence just as a distinct value that comes out of the sequence. (So this makes both models more pure. You can represent the whole thing as just functions at that point—you don't need interfaces.)

And I think "single responsibility" is not the right abstraction here for two reasons. First, the three `IObserver<T>` methods are tightly coupled: their joint usage is governed by a grammar, so none of these methods is used in isolation. Second, these are literally an alternative expression of what `IEnumerable<T>` and `IEnumerator<T>` signify, and the fact that these three aspects are not fully separated there means that they aren't truly separable in `IObserver<T>`. There may be three distinct methods but they are connected.

So if I were to attempt to express what I think you're getting at, I'd say it's that being push-based makes the possibility of exceptions an explicit part of the model. (The explicit separation between "next value" and "end" was in fact already present in `IEnumerable<T>` with the split between `Current` and `MoveNext`). It forces us to see something that was already there but which was much easier to overlook in the `IEnumerable<T>` model. But equally, there are facets of the abstraction which are inherent in the `IEnumerable<T>` construction that are not directly visible in the Rx formulation, which is why we need to describe the grammar in order to have a complete picture of what `IObserver<T>` is. There are some aspects of that grammar that are simply inherent to the type in the `IEnumerable<T>` formulation.

So I think the moral of the story here is that the .NET type system is not powerful enough to fully express everything about the relationship between the provider and consumer of a sequence. Both `IEnumerable/IEnumerator<T>` and `IObservable/IObserver<T>` capture only their own subsets of the whole story. Different bits of that provider/consumer relationship are explicit and implicit in these two formulations. Neither is really more or less pure or overloaded than the other. Each collapses different parts of the whole.

But I'm not sure the introduction is the place for this sort of type system philosophy. I would expect it to confuse more than it would help someone trying to get their head around this for the first time. Possibly it could go in the appendix on the mathematical underpinnings, although this is more "musings" than "foundations".